### PR TITLE
feat: support fallback probes when a probe attach fails

### PR DESCRIPTION
**Pull request by CameronHall**
Summary: Support fallback probes when a probe attach fails

This should address the issue mentioned in https://github.com/pixie-io/pixie/issues/932. I ran into this while trying to run Stirling on WSL.

Type of change: /kind bug

Test Plan: TBD

### CODE REVIEW

File: src/stirling/bpf_tools/bcc_wrapper.cc
Comment (ddelnano) [inline src/stirling/bpf_tools/bcc_wrapper.cc:185]: @CameronHall while we wait for approval to run the GitHub actions, there is a missing comma here that will fix the test build.

File: src/stirling/source_connectors/socket_tracer/socket_trace_connector.cc
Comment (ddelnano) [inline src/stirling/source_connectors/socket_tracer/socket_trace_connector.cc:341]: There are also some linting errors here that can be identified by running `arc lint` (assuming your development environment is setup).

**Review by ddelnano — COMMENTED** [review]
Thanks for the contribution @CameronHall! A few minor comments, but this approach lgtm for handling this fallback behavior.

**Comment by CameronHall** [comment]
Thanks @ddelnano. I've addressed the linting issues

File: src/stirling/bpf_tools/probe_specs/probe_specs.h
Comment (JamesMBartlett) [inline src/stirling/bpf_tools/probe_specs/probe_specs.h:79]: This should be a managed ptr (`std::unique_ptr`) otherwise the memory will be leaked when the struct is deallocated.

**Review by JamesMBartlett — COMMENTED** [review]
Sorry for the delay in reviewing. One minor change and then we should be able to get this through.

**Comment by CameronHall** [comment]
@JamesMBartlett sorry it took so long. Updated it now.

File: src/stirling/bpf_tools/probe_specs/probe_specs.h
Comment (JamesMBartlett) [inline src/stirling/bpf_tools/probe_specs/probe_specs.h:79]: Since this is still a pointer to a struct you will still leak memory. What I was imagining with my comment is:
```
std::unique_ptr<KProbeSpec> fallback_probe = nullptr;
```

File: src/stirling/source_connectors/socket_tracer/socket_trace_connector.cc
Comment (JamesMBartlett) [inline src/stirling/source_connectors/socket_tracer/socket_trace_connector.cc:337]: And then over here you can do `std::make_unique<bpf_tools::KProbeSpec>(...)`

File: src/stirling/source_connectors/socket_tracer/socket_trace_connector.cc
Comment (JamesMBartlett) [inline src/stirling/source_connectors/socket_tracer/socket_trace_connector.cc:341]: ditto

**Review by JamesMBartlett — COMMENTED** [review]
Sorry for the delay on my end and the misunderstanding with my previous comment. Hopefully, I described what I was thinking a little more clearly this time.

**Comment by ddelnano** [comment]
@CameronHall thanks again for getting this PR open. There was recent interest in this change from a community slack discussion, so I went ahead and added on top of your change to address the remaining feedback.

@JamesMBartlett `std::unique_ptr` runs into problems since [this](https://github.com/pixie-io/pixie/blob/79886a48cd2f8a1b77a57b2a314dfe708454f8ce/src/stirling/bpf_tools/bcc_wrapper.cc#L190) relies on a copy constructor being present. Instead of trying to working around that, I chose to use a `std::shared_ptr`. Let me know what you think.

**Comment by JamesMBartlett** [comment]
> @JamesMBartlett `std::unique_ptr` runs into problems since [this](https://github.com/pixie-io/pixie/blob/79886a48cd2f8a1b77a57b2a314dfe708454f8ce/src/stirling/bpf_tools/bcc_wrapper.cc#L190) relies on a copy constructor being present. Instead of trying to working around that, I chose to use a `std::shared_ptr`. Let me know what you think.

I think we should reserve using shared pointers for when the memory is actually shared between two different owners (in this case its just allocated at init), so IMO we should just change the usage in BCCWrapper. Looking at it, it seems that the probe specs are only ever used as a const reference so we should just change the class to store const references instead of copying the probe spec.

**Comment by ddelnano** [comment]
> it seems that the probe specs are only ever used as a const reference so we should just change the class to store const references instead of copying the probe spec.

@JamesMBartlett are you talking about changing the `kprobes_` member to that data type? I believe `std::vector` elements need to be assignable and after trying that I haven't been able to find an option that compiles. I feel like `std::vector` is probably not the right fit and a read only container would make it easier to accomplish what you described.

If that wasn't what you were thinking, let me know and I'm happy to try it.

**Review by JamesMBartlett — APPROVED** [review]
Discussed offline. Decided that its easier to just go with the shared ptr.

**Pull request merged**
